sstable: add parallelized RewriteKeySuffixes function

dt opened the pull request:
This adds a function to the sstable package that can replace some suffix
with another suffix from every key in an sstable, assuming enforcing
that the sstable consists of only Sets.

If the replacement is passed a filter policy, rather than build new
filters, the filter blocks from the original sstable are copied over,
but only after checking that it used the same filter policy and the same
splitter, and that that splitter splits the same length that as is being
replaced i.e. that the pre-replacement filter remains valid
post-replacement.

Data blocks are rewritten in parallel. This informs a few choices in how
it does so: rewritten keys are not passed to `addPoint`, but rather are
directly handed to a worker-local blockWriter. Finished blocks are not
immediately flushed to the output or added to the index, but rather are
added to a buffer, which is then iterated and flushed at the end.

A significant implication of not passing each key to `addPoint` is that
individual KVs are not passed to the table and block property collectors
during rewriting. Instead, when flushing the finished blocks, each table
collector is passed just the first key of the table and each block prop
collector is passed the first key of each block. For the collectors that
CockroachDB uses, to which only the timestamp key suffix is relevant,
this approach may be sufficient, as during replacement all keys in all
blocks have the same suffix. However extending the collector API could
allow this to be more general, e.g. either by making a separate API for
this approach, like AddOneKeyForBlock, or by making the collectors able
to collect and aggregate partial results, so each rewriter could have a
local partial collector and then merge their results. However as the
simple approach of just passing one key per block to the existing API
appears sufficient for CockroachDB, this is left for later exploration.

Benchmarking this implementation compared to simply opening a writer and
feeding each key from a reader to it shows modest gains even without the
addition of concurrency, just from skipping rebuilding filters and props
from every key, with a 12M SST rewriting at about 33% faster. With the
addition of concurrency=8, this jumps to about a 7.5x speedup over the
naive read/write iteration, of 5x over concurrency=1.

```
BenchmarkRewriter
    writer_test.go:357: rewriting a 12 M SSTable
BenchmarkRewriter/RewriteKeySuffixes,concurrency=1
BenchmarkRewriter/RewriteKeySuffixes,concurrency=1-10                130          91251942 ns/op
BenchmarkRewriter/RewriteKeySuffixes,concurrency=8
BenchmarkRewriter/RewriteKeySuffixes,concurrency=8-10                666          18205387 ns/op
BenchmarkRewriter/read-and-write-new-sst
BenchmarkRewriter/read-and-write-new-sst-10                           88         134926796 ns/op
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/pebble/1377)
<!-- Reviewable:end -->

jbowens reviewed (commented):
Nice!

We do anticipate computing block property collectors over data beyond just the key's timestamp. We've been considering explorations around pushing down SQL filters and projections into Pebble, and using block property collectors to aggregate statistics over keys within a block.

Concurrency while reading/decompressing and writing/compressing is desirable more generally (eg, #599), and I'm hopeful that we can combine these use cases into one facility that offers parallelism. I suspect that we can get the performance we need while doing a sequential iteration that computes block property collectors sequentially but defers everything else to parallel workers.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-Mons2EDBi3N8WK2nlqk:bz41ez6)* status: 0 of 7 files reviewed, all discussions resolved (waiting on @erikgrinaker and @sumeerbhola)



<!-- Sent from Reviewable.io -->

dt commented:
@jbowens I'm not _too_ worried about future table and block prop collectors that require seeing every kv.  I think for table prop collectors we could extend their API to have a `AddAll(other PropCollector)` function s.t. each worker could have a local collector them merge them at the end. For block collectors, I think we could probably have a local instance of the collector be passed every kv and collect props for a block, then encode its result and store it along side the block itself in the buffer waiting to be flushed, with some tweak to the API to reduce the assumption that a single instance sees all blocks in order for the collector used in the local worker and then potentially having some sum op over all the encoded ones at the end ala the table collectors. Finally, we might also measure the cost of collecting those props during ingestion and determine is simply isn't worth it -- if it is even 10% overhead, that might be enough for us to say that getting RESTORE done sooner is more important since that is our disaster recovery time. Basically, I think we have plenty of options, so I'm not too worried about figuring out that future when it comes, and for the current collectors, the simple thing works fine.

sumeerbhola reviewed (commented):
> Data blocks are rewritten in parallel. 

Do you have a cpu profile of this benchmark. I am curious what are the big consumers of cpu in the part that has been parallelized.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-MoorLO74lFYck2SV6EU:bwa10p2)* status: 0 of 7 files reviewed, 1 unresolved discussion (waiting on @dt and @erikgrinaker)

___
*[sstable/suffix_rewriter.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MooquAC5U6ahICmXF5O:-MooquAC5U6ahICmXF5P:b-o9gkq7) ([raw file](https://github.com/cockroachdb/pebble/blob/2b3cd36fffb4f6f2af2aa86fb9ec327beb8eb44e/sstable/suffix_rewriter.go#L92)):*
> ```Go
> 	}
> 
> 	for i := worker; i < len(input); i += totalWorkers {
> ```

there is an assumption here that the work per block is not heavily skewed, which is why this static distribution that gives each worker an equal number of blocks is acceptable. Worth noting in a code comment.


<!-- Sent from Reviewable.io -->

dt commented:
> what are the big consumers of cpu in the part that has been parallelized.

I can post an actual profile later, but when I profiled it earlier it looked like the biggest was compression and decompression, with this implementation spending about 21% in compression vs 12% in loop reader/writer and  9% vs 8% in decompression. blockWriter.store was pretty large too IIRC. I looked briefly at allocs and tried buffering the Clone allocs which cut them by >30%, but that yielded no significant top line perf savings.

dt commented:
@jbowens Another thing that occurred to me last night: ideally we'd _copy over_ most props blindly instead of recomputing them, since the whole point is to minimize cost, and perhaps the (enforced) stipulation that suffix replacement _only_ modifies the suffix after Split() is our friend here: we can add something to the prop collector API that says "hey, I'm copying this block but with this new suffix... what's the new prop?" and it could just be identity func for any props that are just on the sql keys / values, to which the suffix is irrelevant. Now I imagine some prop collectors might be interested in the suffix, say, mvcc stats, but there too, I think one of the stipulations of suffix replacement is key: all the keys had the same pre-replacement suffix, so changing it to another one won't change the _relative_ suffix of any two keys in the file, so I'd guess you still don't need any key-by-key recompilation, just maybe an update to prop.

Anyway, this is all speculation until we actually implement any props other than the timestamps, but I'm bullish that given the very narrow definition of suffix replacement (all sets, all with the same suffix, all updates to the same new suffix) that we could make most possible prop collectors work without key-by-key recomputaiton, or if we had to, figure out how aggregate partial parallel computation.

dt commented:
So I tried adding a `addProps` call during the loop to flush produced blocks , which just decompresses it and iterates over it to pass all keys to the prop collectors sequentially. The impact is pretty striking:
```
name                                                         old time/op    new time/op    delta
RewriteSST/keys=1000000/RewriteKeySuffixes,concurrency=8-10    15.0ms ± 3%    39.3ms ± 1%  +161.52%  (p=0.000 n=9+10)

name                                                         old speed      new speed      delta
RewriteSST/keys=1000000/RewriteKeySuffixes,concurrency=8-10   804MB/s ± 3%   307MB/s ± 1%   -61.77%  (p=0.000 n=9+10)

name                                                         old alloc/op   new alloc/op   delta
RewriteSST/keys=1000000/RewriteKeySuffixes,concurrency=8-10    18.5MB ± 0%    18.5MB ± 0%    +0.07%  (p=0.000 n=8+9)

name                                                         old allocs/op  new allocs/op  delta
RewriteSST/keys=1000000/RewriteKeySuffixes,concurrency=8-10     24.6k ± 0%     24.6k ± 0%    +0.02%  (p=0.000 n=7+9)
```

Looking at profile for a run after, it's now just over 20% overall in the `addProps` with that about evenly split between snappy decompression and blockIter.Next.

That looks like enough of a cost to motivate keeping parallel "computation" of block props for now, perhaps with some formalization in the API, and then weighing other options if/when another form of prop requires something different?

jbowens commented:
@dt Yeah, good point. It does seem viable in most use cases to copy over properties that are not dependent on the suffix. Only block properties that depend on the combination of key-value pair _and_ the timestamp would run into difficulties. The block property collector in #1378 is an example of a collector that needs to be computed serially, although practically speaking that particular collector wouldn't ever be used for AddSSTable-rewritten sstables that do not contain multiple MVCC versions of the same key. 

Given that we're also implementing #599 right now, I think we should also explore leveraging that work for this use case to see how far it gets us. If we can avoid it, it'd be nice to avoid maintaining two separate sstable concurrency facilities. It might be the case that the concurrency facilities from #599 plus some of the other optimizations you added here (eg, copying the filter block verbatim) can get us where we need to be.

dt commented:
I doubt just #599 will be sufficient on its own: compression is certainly _a_ big driver of the total time, but only about than a third overall; we're still spending about 20% (now that I removed a bunch of cache.Cache mutex contention) in computing blockWriter.store for example, mostly in counting out bytes for sharedprefixlen, and getting that parallelized 8x is still helping us out here. As speed-of-light test I added a compression=none case to the benchmark, just to see what would happen if somehow #599 were able to magically completely eliminate compression:
```
name                                                                        time/op
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=1-10     79.5ms ± 0%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=2-10     42.3ms ± 1%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=4-10     23.4ms ± 1%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=8-10     15.6ms ± 3%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=16-10    15.7ms ± 1%
RewriteSST/NoCompression/keys=1000000/ReaderWriterLoop-10                      132ms ± 0%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=1-10            84.9ms ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=2-10            44.8ms ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=4-10            24.5ms ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=8-10            16.4ms ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=16-10           16.9ms ± 6%
RewriteSST/Snappy/keys=1000000/ReaderWriterLoop-10                             141ms ± 0%

name                                                                        speed
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=1-10    600MB/s ± 0%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=2-10   1.13GB/s ± 1%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=4-10   2.04GB/s ± 1%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=8-10   3.05GB/s ± 3%
RewriteSST/NoCompression/keys=1000000/RewriteKeySuffixes,concurrency=16-10  3.05GB/s ± 1%
RewriteSST/NoCompression/keys=1000000/ReaderWriterLoop-10                    361MB/s ± 0%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=1-10           142MB/s ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=2-10           269MB/s ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=4-10           493MB/s ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=8-10           738MB/s ± 1%
RewriteSST/Snappy/keys=1000000/RewriteKeySuffixes,concurrency=16-10          716MB/s ± 5%
RewriteSST/Snappy/keys=1000000/ReaderWriterLoop-10                          85.8MB/s ± 0%
```

dt commented:
I played with the prop collectors ideas discussed above a little, and I think I found something I'm not wholly opposed to. Code is over on https://github.com/dt/pebble/compare/sst-rewrite...sst-rewrite-prop-collectors?expand=1

TL;DR: an optional extension to the collector interface that a collector can choose to implement if it wants to be usable during replacement, which is one new method that is called for each block with the old prop value for that block and the change in suffix. Technically none of cockroach's current props actually need the old value, since they can derive their state from just from the suffix, so I suppose it is a little bit of wasted extra work to read, alloc and plumb it to them, but I think the increased generality of the API justifies it, and I'm pretty sure that API would support a collector like the one in  #1378 (since just updating the suffix can't change whether or not the prefix change is different, so it could elect to just copy its prior value).

dt commented:
@jbowens Do you think it'd be easier to review and land this as-is then follow up with block property handling, or pull in the block property handling now and land it all at once?

dt commented:
Keeping this rebased over changes like #1442 for our scale tests tests is a little tricky, so it'd be nice to move this one forward, and then I think I just have two more changes queued up after for review (adding block property handling and then avoiding cache contention and copies by reading the underlying block from the memFile.)

jbowens reviewed (commented):
> Keeping this rebased over changes like #1442 for our scale tests tests is a little tricky

That speaks to the ongoing maintenance burden of having an additional code path intertwined with low-level sstable writing logic. :/

I'm still wary such a highly specialized, low-level sstable rountine.  I would love it if we can find the right low-level interfaces such that more of the nitty gritty can be shared between the Writer and `RewriteKeySuffixes`. I'll keep thinking about it.

> Reviewed 1 of 14 files at r13, 1 of 5 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-MtynpQ9-kcoLg5Ort67:b-uypir7)* status: 1 of 14 files reviewed, 3 unresolved discussions (waiting on @dt, @erikgrinaker, and @jbowens)

___
*[sstable/suffix_rewriter.go, line 68 at r16](https://reviewable.io/reviews/cockroachdb/pebble/1377#-Mtyn4pY7QtlmCq5SVTL:-Mtyn4pY7QtlmCq5SVTM:blxzcto) ([raw file](https://github.com/cockroachdb/pebble/blob/e2d67998efcf3ed0e3b6a9b3d055da8b3f02cb10/sstable/suffix_rewriter.go#L68)):*
> ```Go
> 	from, to []byte,
> 	concurrency int,
> ) (*WriterMetadata, error) {
> ```

Re my other comment about narrowing "suffix," we should error if `r.Split` or `o.Comparer.Split` is `nil`.
___
*[sstable/suffix_rewriter.go, line 262 at r16](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MtylsgCDTWubdrA25Bd:-MtylsgDEktiiBC1IquD:b-mk2ihl) ([raw file](https://github.com/cockroachdb/pebble/blob/e2d67998efcf3ed0e3b6a9b3d055da8b3f02cb10/sstable/suffix_rewriter.go#L262)):*
> ```Go
> 				return errBadKind
> 			}
> 			if !bytes.HasSuffix(iter.Key().UserKey, from) {
> ```

I think we should restrict `RewriteKeySuffixes` to replacing the Pebble notion of "suffix": It's confusing to have multiple concepts of 'suffixes',  we don't need anything more general, and replacing more than the Pebble notion of "suffix" is unsupported if `r.Split != nil`. So this should be:

```
uk := item.Key().UserKey
si := r.Split(uk)
if !bytes.Equal(uk[si:], from) {
```


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377)* status: 0 of 14 files reviewed, 3 unresolved discussions (waiting on @dt, @erikgrinaker, and @jbowens)

___
*[sstable/suffix_rewriter.go, line 68 at r16](https://reviewable.io/reviews/cockroachdb/pebble/1377#-Mtyn4pY7QtlmCq5SVTL:-Mtz-8m_8wpMn5dWG-YT:b-o9o6ik) ([raw file](https://github.com/cockroachdb/pebble/blob/e2d67998efcf3ed0e3b6a9b3d055da8b3f02cb10/sstable/suffix_rewriter.go#L68)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Re my other comment about narrowing "suffix," we should error if `r.Split` or `o.Comparer.Split` is `nil`.
</blockquote></details>

Added check of `o.Comparer` and `o.Comparer.Split`. Do we actually care about `r.Split`? Any reason not to just use just use o.Comparer.Split throughout?
___
*[sstable/suffix_rewriter.go, line 262 at r16](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MtylsgCDTWubdrA25Bd:-MtyxKiC3jXalB3CqC00:bjlbrc5) ([raw file](https://github.com/cockroachdb/pebble/blob/e2d67998efcf3ed0e3b6a9b3d055da8b3f02cb10/sstable/suffix_rewriter.go#L262)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

I think we should restrict `RewriteKeySuffixes` to replacing the Pebble notion of "suffix": It's confusing to have multiple concepts of 'suffixes',  we don't need anything more general, and replacing more than the Pebble notion of "suffix" is unsupported if `r.Split != nil`. So this should be:

```
uk := item.Key().UserKey
si := r.Split(uk)
if !bytes.Equal(uk[si:], from) {
```
</blockquote></details>

Hm, yeah. I'd previously been _mostly_ enforcing that we had only one notion of suffix by checking, down below, that the suffix according to w.split of the first block's sep key was `from` -- combined with asserting here that every key had the _byte_ suffix `from` that would imply it had the same `split()` suffix as well when said split is fixed length or based on the suffix content (like CockroachComparer), but I suppose that might not hold for a different Split func (say, var-int prefix). Switched to calling Split() on every key now instead as suggested.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
drive-by comments -- feel free to ignore

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-MuBukWl6fXZqXJc8BWW:bajp488)* status: 0 of 14 files reviewed, 6 unresolved discussions (waiting on @dt, @erikgrinaker, and @jbowens)

___
*[sstable/suffix_rewriter.go, line 20 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsBvU1AdwuvJV265V:-MuBsBvU1AdwuvJV265W:begia79) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L20)):*
> ```Go
> // required that the SST consist of only SETs, and that every key have `from` as
> // its suffix, and furthermore the length of from must match the suffix length
> // designated by the Split() function if one is set in the WriterOptions.
> ```

is the exception `if one is set in the WriterOptions` no longer relevant since we are enforcing that suffix is the one returned by `Split`?
___
*[sstable/suffix_rewriter.go, line 37 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsXkz7jbYPo_DDzBT:-MuBsXkz7jbYPo_DDzBU:b-r3q878) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L37)):*
> ```Go
> //
> // 1) to be able to copy unmodified filters, it is required that the writer's
> // comparer have a Split function and that the replaced suffix of each key not
> ```

this would not need to be repeated here, if this is a general requirement.
___
*[sstable/suffix_rewriter.go, line 43 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsvXLBAQ3hBWJXRpw:-MuBsvXLBAQ3hBWJXRpx:bjzzfko) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L43)):*
> ```Go
> // 2) any block and table property collectors in-use must be able to derive a
> // correct result -- for a block wherein *all keys had and will have the same
> // suffix* -- from a single updated key per block.
> ```

how important is this optimization in terms of performance numbers? I ask since unlike the other restrictions specified here, this one (subjectively) feels like a hack. 


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> That speaks to the ongoing maintenance burden of having an additional code path intertwined with low-level sstable writing logic.

Eh, I mean, sure, but I think it's cheaper to update it all at once than it is to go back during a rebase and figure out where "w.restartInterval" went. As far as figuring out the common shared APIs rather than intertwining, I imagine the answer is to merge and then start to see where to draw the lines e.g. would rewriter and writer be happy to share a common `indexWriter` helper struct, instead of rewriter having a Writer which it uses just for flushing indexes? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-MtzuMpR8MtoC_HDB3zt:bbryqr)* status: 0 of 14 files reviewed, 6 unresolved discussions (waiting on @dt, @erikgrinaker, @jbowens, and @sumeerbhola)

___
*[sstable/suffix_rewriter.go, line 20 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsBvU1AdwuvJV265V:-MuC7ojO91aHGEqhLPPM:b-m8eat5) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L20)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is the exception `if one is set in the WriterOptions` no longer relevant since we are enforcing that suffix is the one returned by `Split`?
</blockquote></details>

Yeah, correct, we're requiring WriterOptions.Comparer.Split be set to call this at all. Will update.
___
*[sstable/suffix_rewriter.go, line 37 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsXkz7jbYPo_DDzBT:-MuC7mW1-2WrWmn0ICaB:bu9ggeo) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L37)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

this would not need to be repeated here, if this is a general requirement.
</blockquote></details>

Yeah, this is out of date. Will fix.
___
*[sstable/suffix_rewriter.go, line 43 at r19](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuBsvXLBAQ3hBWJXRpw:-MuC7cpQ5_o5j3qLWXwe:b-tptdu6) ([raw file](https://github.com/cockroachdb/pebble/blob/19eb26e7b44fcc19bcc303545a086d68b1f9347d/sstable/suffix_rewriter.go#L43)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how important is this optimization in terms of performance numbers? I ask since unlike the other restrictions specified here, this one (subjectively) feels like a hack. 
</blockquote></details>

I opined about this hack (and agree completely that that is what it is) above, and linked to a change that I have queued up to follow this that I think improves on it significantly with an extension to the prop collector API: https://github.com/dt/pebble/commit/42070680d60606f515edb5ff01096a7132e52e3f


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377)* status: 0 of 14 files reviewed, 6 unresolved discussions (waiting on @erikgrinaker, @jbowens, and @sumeerbhola)

___
*[sstable/suffix_rewriter.go, line 92 at r5](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MooquAC5U6ahICmXF5O:-MuCqSYY0Idfq0vkjW1v:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/2b3cd36fffb4f6f2af2aa86fb9ec327beb8eb44e/sstable/suffix_rewriter.go#L92)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

there is an assumption here that the work per block is not heavily skewed, which is why this static distribution that gives each worker an equal number of blocks is acceptable. Worth noting in a code comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jbowens reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 4 files at r18, 1 of 1 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377#-:-MuHjiRNElJvoURfgTWi:bnfp4nl)* status: 3 of 14 files reviewed, 13 unresolved discussions (waiting on @dt, @erikgrinaker, @jbowens, and @sumeerbhola)

___
*[sstable/suffix_rewriter.go, line 20 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHLS2k2RLZ8797XFXE:-MuHLS2k2RLZ8797XFXF:b-lo43yh) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L20)):*
> -- the   an sstable to `out` containing  the sstable in

looks like some of this comment got mangled
___
*[sstable/suffix_rewriter.go, line 26 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHM9Zu7EpGQsxXPMZT:-MuHM9Zu7EpGQsxXPMZU:bmptipw) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L26)):*
> ```Go
> // simply passed each key from the Reader, and its result is returned. In this
> // implementation, all filters and properties are re-derived, just as if the new
> // SST had been constructed from arbitrary external keys.
> ```

Could we just use the one, concurrent implementation with `concurrency=1` if `concurrency` is set to zero? It seems preferable to just have the one implementation.
___
*[sstable/suffix_rewriter.go, line 141 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHNd7RBWOktNb7QD3D:-MuHNd7RBWOktNb7QD3E:b-wjynbu) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L141)):*
> ```Go
> 	}
> 
> 	// Copy over the filter block if it exists (if it does, copyDataBlocks will
> ```

nit: s/copyDataBlocks/rewriteDataBlocksToWriter/  ?
___
*[sstable/suffix_rewriter.go, line 181 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHfzF720fN4mDlpCHI:-MuHfzF720fN4mDlpCHJ:b6sgco9) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L181)):*
> ```Go
> 	errCh chan error,
> 	g *sync.WaitGroup,
> ) {
> ```

nit: this is a pretty thin function with many parameters. inline in an anonymous function in the call site?
___
*[sstable/suffix_rewriter.go, line 201 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHignH2yS8DkLn62pq:-MuHignH2yS8DkLn62pr:b-o57tc6) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L201)):*
> ```Go
> 		restartInterval: restartInterval,
> 	}
> 	var compressionBuf []byte
> ```

can this use the new `dataBlockBuf`?
___
*[sstable/suffix_rewriter.go, line 249 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHgzHB8kwSGw5_iCyg:-MuHgzHC7w2BdoDQafmS:b-cc14ym) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L249)):*
> ```Go
> 				return err
> 			}
> 			prefixLen := len(key.UserKey) - len(from)
> ```

nit: can use `si` instead of computing from `len(key.UserKey) - len(from)`
___
*[sstable/suffix_rewriter.go, line 305 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHiYmMDavRPk4g0kPR:-MuHiYmMDavRPk4g0kPS:bv5d6qp) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L305)):*
> ```Go
> 		default:
> 			return errors.Newf("unsupported checksum type: %d", checksumType)
> 		}
> ```

this can be a followup, but this can now use `checksummer.checksum`.
___
*[sstable/suffix_rewriter.go, line 349 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHjArVARX2qpI79vOz:-MuHjArVARX2qpI79vP-:b-ktvgnn) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L349)):*
> ```Go
> 	for i := range blocks {
> 		// Write the rewritten block to the file.
> 		n, err := w.writer.Write(blocks[i].data)
> ```

This can be a followup, if at all, but this can be done in parallel to individual block rewriting with additional synchronization, right? Do you know how much time is spent in the above block rewriting versus this loop, copying into the memfile and constructing the index?


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377)* status: 3 of 14 files reviewed, 11 unresolved discussions (waiting on @dt, @erikgrinaker, @jbowens, and @sumeerbhola)

___
*[sstable/suffix_rewriter.go, line 26 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHM9Zu7EpGQsxXPMZT:-MuHqgBw-HhGG923zSNb:b-t7xa2u) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L26)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

Could we just use the one, concurrent implementation with `concurrency=1` if `concurrency` is set to zero? It seems preferable to just have the one implementation.
</blockquote></details>

the 0 vs 1 implementation was mostly for ease of testing. I could just make two separate functions though and then let the caller call the right one. In that case, I think concurrency < 1 is an error.
___
*[sstable/suffix_rewriter.go, line 141 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHNd7RBWOktNb7QD3D:-MuIpGLF4evITbkiDl_U:bniq3l9) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L141)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: s/copyDataBlocks/rewriteDataBlocksToWriter/  ?
</blockquote></details>

Ah, yep, thanks.
___
*[sstable/suffix_rewriter.go, line 181 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHfzF720fN4mDlpCHI:-MuIpJsd2gbn27NUYqfp:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L181)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: this is a pretty thin function with many parameters. inline in an anonymous function in the call site?
</blockquote></details>

Done.
___
*[sstable/suffix_rewriter.go, line 201 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHignH2yS8DkLn62pq:-MuMXOd54UhAEPAbJbQr:b-r69xu3) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L201)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

can this use the new `dataBlockBuf`?
</blockquote></details>

Switched to blockBuf. 
___
*[sstable/suffix_rewriter.go, line 249 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHgzHB8kwSGw5_iCyg:-MuJ88JH7YGP15TSaO6p:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L249)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

nit: can use `si` instead of computing from `len(key.UserKey) - len(from)`
</blockquote></details>

Done.
___
*[sstable/suffix_rewriter.go, line 305 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHiYmMDavRPk4g0kPR:-MuJKQvO6JuWAcHFHn9X:bplu8ia) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L305)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

this can be a followup, but this can now use `checksummer.checksum`.
</blockquote></details>

`compressAndChecksum`, right? this removes a bunch of the repeated code here, so that seems like a win. Unfortunately seems to force moving blockBuf to a pointer so I pick up another alloc but I'll switch it for now and then golf the perf a bit later if it is needed.
___
*[sstable/suffix_rewriter.go, line 349 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHjArVARX2qpI79vOz:-MuHpfvL7R4GX2wgQZwQ:b-u7rl7u) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L349)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

This can be a followup, if at all, but this can be done in parallel to individual block rewriting with additional synchronization, right? Do you know how much time is spent in the above block rewriting versus this loop, copying into the memfile and constructing the index?
</blockquote></details>

yeah, we could stat another flusher worker and then have some additional synchronization like atomic's for low watermark of each worker to determine what's flushable or something, I just didn't want to do it now.

Profile wise, don't think these memcopies really showed up last I checked -- IIRC it was mostly moving the bytes around during rewriting.


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1377)* status: 1 of 14 files reviewed, 11 unresolved discussions (waiting on @erikgrinaker, @jbowens, and @sumeerbhola)

___
*[sstable/suffix_rewriter.go, line 20 at r21](https://reviewable.io/reviews/cockroachdb/pebble/1377#-MuHLS2k2RLZ8797XFXE:-MuMYfez6HlJ1_FFAIq_:b-xxajvu) ([raw file](https://github.com/cockroachdb/pebble/blob/9e21d1228cbfa9361821ca290b853e291103cca4/sstable/suffix_rewriter.go#L20)):*
<details><summary><i>Previously, jbowens (Jackson Owens) wrote…</i></summary><blockquote>

> -- the   an sstable to `out` containing  the sstable in

looks like some of this comment got mangled
</blockquote></details>

Reworked it entirely, to remove the whole this-impl-that-impl thing.


<!-- Sent from Reviewable.io -->

dt commented:
@sumeerbhola are you okay with merging this now with the hacky handling of prop collectors and then immediately opening a followup PR specifically for handling them better such as what I linked above? or would you rather pull that change in here first and merge as one when we hash it out?

erikgrinaker commented:
Can we drop the `from` parameter? The current KV API doesn't require input SSTs to have uniform timestamps, and it seems like an arbitrary requirement here.

dt commented:
I think having a `from` parameter will be helpful in the API for updating block and table properties -- if I start thinking about hypothetical property collectors like mvcc stats, knowing the _change_ being applied to the suffix seems to open up more cases where you could update your property than just knowing the new suffix. The KV API also will effectively always get the from parameter to avoid rewrites entirely outside pushes so it only being able to use the fast rewriter when we have it isn't a big restriction for us?

erikgrinaker commented:
>The KV API also will effectively always get the from parameter to avoid rewrites entirely outside pushes so it only being able to use the fast rewriter when we have it isn't a big restriction for us?

Possibly, unless there are cases where we want to ingest pre-built SSTs that we don't want to bother with rewriting before submitting?

If we want to make this a hard requirement I'm going to have to shuffle some stuff around, but nbd.

dt commented:
I think we can keep things as they are for now and only use the accelerated rewriter if SSTTimestamp is set and revisit if/when anybody starts calling AddSSTable with WriteAtRequestTimestamp but not SSTTimestamp

erikgrinaker commented:
Yeah, let's see when we integrate this into KV. Easier to change the API around now before we have any callers, so if we think it's always going to be set we may as well design around that.

dt commented:
I'm happy to revisit the `from` matching requirement once we sort out the block and table props API as well, if we decide it knowing the delta isn't actually required there.

dt commented:
Is there anything left to resolve on this one before moving on to the follow-ups for property collectors?

sumeerbhola commented:
> @sumeerbhola are you okay with merging this now with the hacky handling of prop collectors and then immediately opening a followup PR specifically for handling them better such as what I linked above? or would you rather pull that change in here first and merge as one when we hash it out?

hacky handling is fine for this merge.

dt commented:
Thanks for the reviews!
